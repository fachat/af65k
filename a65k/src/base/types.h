/****************************************************************************

    type definitions
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef TYPES_H
#define TYPES_H

#include <stdbool.h>

typedef struct _type type_t;

// more or less a class definition type
struct _type {
	const char 	*name;
	int		sizeoftype;
};

typedef bool 	bool_t;

// because the standard isprint() does not handle UTF8 codes
static inline bool_t isPrint(char c) {
	return c > 31;
}

#endif

