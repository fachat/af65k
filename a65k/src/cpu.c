
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <string.h>

#include "mem.h"
#include "hashmap.h"
#include "astring.h"
#include "cpu.h"

#include "infiles.h"
#include "errors.h"


static const cpu_t cpus[] = {
	{ CPU_NMOS,
		"nmos",			ISA_NMOS_ALL,	
		W_WORD,	true, 	false,
		"Standard NMOS 6502, with Decimal mode and illegal opcodes." },
	{ CPU_NMOS_NO_ILLEGAL,
		"nmos_no_illegal",	ISA_NMOS + ISA_NMOS_BCD,	
		W_WORD,	true, 	true,
		"Standard NMOS 6502 with decimal mode but without illegal opcodes." },
	{ CPU_CMOS,
		"cmos",			ISA_NMOS + ISA_CMOS,	
		W_WORD,	true, 	false,
		"Standard 65C02 CMOS, without the Rockwell extensions." },
	{ CPU_RCMOS,
		"cmos_rockwell",	ISA_NMOS + ISA_NMOS_BCD + ISA_CMOS,	
		W_WORD,	true, 	false,
		"R65C02 Rockwell CMOS with the RMB/SMB/BBR/BBS extensions." },
	{ CPU_NO_BCD,
		"nmos_no_bcd",		ISA_NMOS + ISA_NMOS_ILLEGAL,	
		W_WORD,	false, 	false,
		"NMOS 6502 with illegal opcodes, but no decimal mode (SED)." },
	{ CPU_NO_BCD_NO_ILLEGAL,
		"nmos_no_illegal_no_bcd", ISA_NMOS,	
		W_WORD,	false,	true,
		"NMOS 6502 without illegal opcodes, and no decimal mode." },
	{ CPU_802,
		"65802",		ISA_816_ALL,
		W_816,	true,	false,
		"WDC65802" },
	{ CPU_816,
		"65816",		ISA_65K_ALL,	
		W_816,	true, 	false,
		"WDC65816" },
	{ CPU_65K,
		"65k",			ISA_65K_ALL,	
		W_QUAD,	true, 	false,
		"Generic 65K CPU" },
	{ CPU_65K_W,
		"65k_w",		ISA_65K_ALL,	
		W_WORD,	true, 	false,
		"65K CPU in 16 bit width." },
	{ CPU_65K_L,
		"65k_l",		ISA_65K_ALL,	
		W_LONG,	true, 	false,
		"65K CPU in 32 bit width." },
	{ CPU_65K_Q,
		"65k_q",		ISA_65K_ALL,	
		W_QUAD,	true, 	false,
		"65K CPU in 64 bit width." },
};

static hash_t *cpu_typemap = NULL;

static const void *cpu_key_from_entry(const void *entry) {
	cpu_t *cpu = (cpu_t*) entry;
	return cpu->name;
}

void cpu_module_init() {

	int num_cpus = sizeof(cpus)/sizeof(cpu_t);

	cpu_typemap = hash_init_stringkey(10, 5, cpu_key_from_entry);

	for (int i = 0; i < num_cpus; i++) {
		hash_put(cpu_typemap, (void*)&cpus[i]);
	}
}

const cpu_t *cpu_by_name(const char *name) {

	cpu_t *cpu = (cpu_t*) hash_get(cpu_typemap, name);

	return cpu;
}

const cpu_t *cpu_by_type(cpu_type type) {

	unsigned int num_cpus = sizeof(cpus)/sizeof(cpu_t);
	if (type >= num_cpus) {
		error_illegal_cpu_type(type, num_cpus);
		return NULL;
	}
	return &cpus[type];
}



