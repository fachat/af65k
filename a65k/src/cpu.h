
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef CPU_H
#define CPU_H

typedef enum {
        W_NONE                  = 0,
        W_8                     = 1,
        W_16                    = 2,
        W_24                    = 4,
        W_32                    = 8,
        W_64                    = 16,
} width_map;

// possible register widths (i.e. W_24 3-byte addresses are missing)
#define W_ALLREG        (W_8 + W_16 + W_32 + W_64)
#define W_816           (W_8 + W_16 + W_24)
#define W_BYTE          (W_8)
#define W_WORD          (W_8 + W_16)
#define W_LONG          (W_8 + W_16 + W_32)
#define W_QUAD          (W_8 + W_16 + W_32 + W_64)



typedef enum {
        ISA_NMOS                = 1,    // standard NMOS without SED
        ISA_NMOS_ILLEGAL        = 2,    // NMOS illegal opcodes
        ISA_NMOS_BCD            = 4,    // NMOS SED opcode
        ISA_CMOS                = 8,    // CMOS extensions
        ISA_CMOS_ROCKWELL       = 16,   // RMB/SMB/BBS/BBR opcodes
        ISA_816                 = 32,   // 65816 extensions to CMOS 6502
        ISA_65K                 = 64,   // 65K extensions to CMOS 6502
} isa_map;

#define ISA_NMOS_ALL    (ISA_NMOS + ISA_NMOS_ILLEGAL + ISA_NMOS_BCD)
#define ISA_CMOS_ALL    (ISA_NMOS + ISA_CMOS + ISA_CMOS_ROCKWELL)
#define ISA_816_ALL     (ISA_NMOS + ISA_CMOS + ISA_816)
#define ISA_65K_ALL     (ISA_NMOS + ISA_CMOS + ISA_65K)
#define ISA_ALL         (ISA_NMOS + ISA_NMOS_ILLEGAL + ISA_NMOS_BCD + ISA_CMOS + ISA_CMOS_ROCKWELL + ISA_816 + ISA_65K)


typedef enum {
	CPU_NMOS		= 0,	// legal NMOS opcodes, decimal mode and illegal opcodes
	CPU_NMOS_NO_ILLEGAL	= 1,	// legal NMOS opcodes incl. decimal mode, but no illegal opcodes
	CPU_CMOS		= 2,	// CMOS opcodes (without Rockwell extensions)
	CPU_RCMOS		= 3,	// CMOS plus Rockwell extensions
	CPU_NO_BCD		= 4,	// NMOS with illegal opcodes, but without BCD mode (e.g. NES)
	CPU_NO_BCD_NO_ILLEGAL	= 5,	// NMOS without illegal opcodes, and without BCD mode
	CPU_802			= 6,	// 65802
	CPU_816			= 7,	// 65816
	CPU_65K			= 8,	// base 65k CPU (implicitely 64 bit, modifyable with .width) 
	CPU_65K_W		= 9,	// 65k CPU 16-bit width
	CPU_65K_L		= 10,	// 65k CPU 32-bit width
	CPU_65K_Q		= 11,	// 65k CPU 64-bit width
} cpu_type;

typedef struct {
	const cpu_type	type;
	const char 	*name;
	const isa_map	isa;
	const width_map	width;		// memory model size 16, 24, 32, 64
	const bool_t	has_bcd;
	const bool_t	has_illegal;
	const char 	*desc;		// description of CPU model
} cpu_t;


void cpu_module_init();

const cpu_t *cpu_by_name(const char *name);

const cpu_t *cpu_by_type(cpu_type type);


#endif

