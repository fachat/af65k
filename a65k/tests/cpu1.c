

#include <stdio.h>

#include "types.h"
#include "position.h"
#include "cpu.h"


int main(int argc, char *argv[]) {

	cpu_module_init();

	const cpu_t *c = cpu_by_name("nmos");

	printf("nmos -> %s, type=%d, width=%d\n",
		c->name, c->type, c->width);


	c = cpu_by_name("rcmos");
	if (c != NULL) {
		printf("found wrong CPU type %s\n", c->name);
	}

	c = cpu_by_name("cmos_rockwell");
	printf("rcmos -> %s, type=%d, width=%d\n",
		c->name, c->type, c->width);

	c = cpu_by_type(NULL, -123);
	if (c != NULL) {
		printf("found wrong CPU type %d\n", -123);
	}

	c = cpu_by_type(NULL, CPU_816);
	printf("816 -> %s, type=%d, width=%d\n",
		c->name, c->type, c->width);

}



