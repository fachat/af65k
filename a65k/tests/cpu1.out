nmos -> nmos, type=0, base=0, width=16, has_bcd=1, has_illegal=0
rcmos -> cmos_rockwell, type=3, base=2, width=16, has_bcd=1, has_illegal=0
816 -> 65816, type=7, base=7, width=24, has_bcd=1, has_illegal=0
